src/antsibull/sanity_tests.py

```diff
@@ -0,0 +1,361 @@
+# Copyright (C) 2023 Maxwell G <maxwell@gtmx.me>
+# SPDX-License-Identifier: GPL-3.0-or-later
+# GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or
+# https://www.gnu.org/licenses/gpl-3.0.txt)
+
+
+"""
+Utilities to run sanity tests accross collections
+"""
+
+from __future__ import annotations
+
+import datetime
+import json
+import shlex
+import shutil
+import sys
+from collections.abc import Collection, Iterable, Iterator, Sequence
+from dataclasses import dataclass
+from functools import partial
+from pathlib import Path
+from subprocess import CompletedProcess
+from typing import TYPE_CHECKING, Any, TypedDict
+
+from antsibull_core import app_context
+from antsibull_core.logging import log
+from antsibull_core.subprocess_util import log_run
+from antsibull_core.yaml import store_yaml_file
+from packaging.version import Version
+
+from antsibull.constants import SANITY_TESTS_BANNED_IGNORES, SANITY_TESTS_DEFAULT
+from antsibull.types import CollectionName, add_dataclass_yaml_type
+
+if TYPE_CHECKING:
+    from _typeshed import StrPath
+
+mlog = log.fields(mod=__name__)
+
+
+class SanityOutput(TypedDict):
+    """
+    Mapping of `ansible-test sanity` output and other related data
+    """
+
+    cmd: list[str]
+    returncode: int
+    stdout: str
+    stderr: str
+    runtime: float
+    test_json: dict[str, Any]
+    ignore_entries: list[IgnoreEntry]
+    banned_ignore_entries: list[IgnoreEntry]
+    ignores_file: str | None
+
+
+class CollectionOutput(TypedDict):
+    """
+    Collection entry
+    """
+
+    failed: bool
+    sanity: SanityOutput
+
+
+class EnvDetails(TypedDict):
+    """
+    Details about the ansible-test environment
+    """
+
+    ansible_test_version: str
+    sanity_tests: list[str]
+
+
+class Output(TypedDict):
+    """
+    Parent mapping of the sanity test data file
+    """
+
+    collections: dict[CollectionName, CollectionOutput]
+    total_runtime: float
+    env_details: EnvDetails
+
+
+@dataclass(frozen=True)
+class IgnoreEntry:
+    """
+    Represents an entry in an `ansible-test sanity` ignores file
+    """
+
+    file: str
+    test: str
+    remainder: str
+
+    def as_str(self) -> str:
+        return f"{self.file} {self.test}{self.remainder}"
+
+
+add_dataclass_yaml_type(IgnoreEntry)
+
+
+def parse_ignores_file(file: StrPath) -> Iterator[IgnoreEntry]:
+    """
+    Parse a sanity test ignore file
+    """
+    flog = mlog.fields(func="parse_sanity_test_file", file=file)
+    with open(file, encoding="utf-8") as fp:
+        for line in fp:
+            line = line.rstrip("\n")
+            parsed = line.split(" ", 2)
```

Comment:
Hmm, `path/to/foo  validate-modules:invalid-documentation` would end up with `parsed[1] == ''`? I'm not 100% whether ansible-test would accept such a line though.

Reply:
I tried

``` diff
diff --git a/tests/sanity/ignore-2.16.txt b/tests/sanity/ignore-2.16.txt
index 12e0b26..1550c79 100644
--- a/tests/sanity/ignore-2.16.txt
+++ b/tests/sanity/ignore-2.16.txt
@@ -1 +1 @@
-plugins/modules/docker_container_copy_into.py validate-modules:undocumented-parameter # _max_file_size_for_diff is used by the action plugin
+plugins/modules/docker_container_copy_into.py  validate-modules:undocumented-parameter # _max_file_size_for_diff is used by the action plugin
```

and `ansible-test sanity` no longer ignored that error after adding the extra space, so it seems this parsing algorithm is correct.